src/rules.rs

```diff
@@ -146,193 +147,105 @@ mod tests {
+        mask = if is_subject {
+            let node = Subject::NamedNode(NamedNode {
+                iri: node_iri.to_string(),
+            });
+            match_type_rule_subject(&node, mask, &type_index, &rules)
+        } else {
+            let node = Term::NamedNode(NamedNode {
+                iri: node_iri.to_string(),
+            });
+            match_type_rule_object(&node, mask, &type_index, &rules)
         };
-        let rules = set_predicate_rule("http://example.org/hasName");
-        let mut mask = TripleMask::default();
-
-        mask = match_predicate_rule(&predicate, mask, &rules);
-
-        assert!(mask.is_set(&TripleMask::OBJECT));
-        assert!(!mask.is_set(&TripleMask::SUBJECT));
-    }
-
-    #[test]
-    fn predicate_not_in_config() {
+        assert_eq!(mask.is_set(&TripleMask::SUBJECT), expected_s);
+        assert_eq!(mask.is_set(&TripleMask::OBJECT), expected_o);
+    }
+
+    #[rstest]
+    // Predicate is in the rules
+    #[case("hasName", "hasName", false, true)]
+    // Predicate is not in the rules
+    #[case("hasName", "hasAge", false, false)]
+    fn predicate_rule(
+        #[case] node_iri: &str,
+        #[case] rule_type: &str,
+        #[case] expected_s: bool,
+        #[case] expected_o: bool,
+    ) {
         let predicate = NamedNode {
-            iri: "http://example.org/hasName".to_string(),
+            iri: node_iri.to_string(),
         };
-        let rules = set_predicate_rule("http://example.org/hasAge");
+        let rules = set_predicate_rule(rule_type);
         let mut mask = TripleMask::default();
 
         mask = match_predicate_rule(&predicate, mask, &rules);
 
-        assert!(!mask.is_set(&TripleMask::OBJECT));
-        assert!(!mask.is_set(&TripleMask::SUBJECT));
-    }
-
-    #[test]
-    fn subject_predicate_in_config() {
-        let alice_iri = "http://example.org/Alice";
-        let person_iri = "http://example.org/Person";
-        let pred_hn = "http://example.org/hasName";
-
+        assert_eq!(mask.is_set(&TripleMask::SUBJECT), expected_s);
+        assert_eq!(mask.is_set(&TripleMask::OBJECT), expected_o);
+    }
+
+    #[rstest]
+    // Subject predicate in config
+    #[case(
+        "Alice", "hasName", "Person", "hasName", "Alice", "Person", false, true
+    )]
+    // Subject in config, predicate not
+    #[case(
+        "Alice", "hasName", "Person", "hasAge", "Alice", "Person", false, false
+    )]
+    // Subject predicate not in config
+    #[case("Alice", "hasName", "Bob", "hasAge", "Alice", "Person", false, false)]
+    // Subject not in type index
+    #[case("Alice", "hasName", "Bob", "hasAge", "Bob", "Person", false, false)]
+    fn subject_predicate_rule(
+        #[case] subject_iri: &str,
+        #[case] predicate_iri: &str,
+        #[case] rule_subject: &str,
+        #[case] rule_predicate: &str,
+        #[case] index_subject: &str,
+        #[case] index_object: &str,
+        #[case] expected_s: bool,
```

Comment:
**suggestion:** when we test the sp rule, we never want to mask the subject (do we?). If so, we could simplify this by removing subject_iri from the arguments and the test cases, while still keeping the assert to check the subject mask.